jep/401/README.adoc

```diff
@@ -0,0 +1,246 @@
+* `preHeader.jelly`: that will provide the content of the elements that want to be provided/rendered before header div is rendered on the browser
+* `headerContent.jelly`: that will provide the content of the header div
+* `postHeader.jelly`: that will provide the content of the elements that want to be provided/rendered after the header div is rendered on the browser
+
+Thus, `pageHeader.jelly` should look like:
+
+```
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
+	<j:set var="header" value="${h.header()}" />
+	<st:include it="${header}" page="headerDocumentation.jelly"/>
+	<st:include it="${header}" page="preHeader.jelly"/>
+	<st:include it="${header}" page="headerContent.jelly"/>
+	<st:include it="${header}" page="postHeader.jelly"/>
+</j:jelly>
+```
+
+The `h.header()` method is provided in the `Functions` and it will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value). In case two or more headers are provided via external plugins with the same priority (and max), then the default one will be provided and a warning message would be provided on the logs.
+
+In terms of simplicity, proposal aims to use only one extension to do a full replacement of the header. As a follow up, each mentioned section of the header could have its own extension to override only a specific part of the header.
+
+=== Extensibility
+
+Jenkins users will be able to interact with the `Header` extension as follows in case want to customize the header of their instances via plugin:
+
+* Extend `Header` and provide an implementation of the priority method giving to this header the max possible value (via ordinal value)
+* Use an `ExtensionFilter` that would make it impossible to provide any additional Header instance
+
+== Motivation
+
+Jenkins uses `pageHeader.jelly` file to specify the content of the header. Although it is valid, if a user wants to modify the header to perform some branding operation, Jenkins header branding capabilities are limited.
+
+There exist some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], that allow Jenkins users to customize some parts like CSS and/or Javascript. On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts. In addition, it the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.
```

Comment:
> then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts

is this too close to internal information? I'd check if it's safe to say that before publishing it.

Reply:
typo (apart from my previous comment)
```suggestion
There exist some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], that allow Jenkins users to customize some parts like CSS and/or Javascript. On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts. In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.
```

> In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration

for me, this is the key point